src/worker/queue.ts

```diff
@@ -48,3 +59,83 @@ export function startQueue(
+                        console.error(error)
+                        Sentry.captureException(error)
+                    }
+                    if (messages?.length) {
+                        console.info(
+                            `🍕 ${messages.length} ${
+                                messages.length === 1 ? 'message' : 'messages'
+                            } consumed from Kafka at ${Date.now()}`
+                        )
+                    } else {
+                        return
+                    }
+                    try {
+                        for (const message of messages) {
+                            const timer = new Date()
+                            const rawEventMessage = JSON.parse(message.value!.toString()) as RawEventMessage
+                            const parsedEventMessage: ParsedEventMessage = {
+                                ...rawEventMessage,
+                                data: JSON.parse(rawEventMessage.data),
+                                now: DateTime.fromISO(rawEventMessage.now),
+                                sent_at: rawEventMessage.sent_at ? DateTime.fromISO(rawEventMessage.sent_at) : null,
+                            }
+                            console.info(`Processing event ${parsedEventMessage.data.event} from WAL`)
+                            const processedEvent = await processEvent({
+                                ...parsedEventMessage,
+                                event: parsedEventMessage.data.event,
+                                properties: parsedEventMessage.data.properties,
+                                now: rawEventMessage.now,
+                                sent_at: rawEventMessage.sent_at,
+                            })
+                            if (processedEvent) {
+                                console.info(`Ingesting event ${parsedEventMessage.data.event}`)
+                                const { distinct_id, ip, site_url, team_id, now, sent_at } = processedEvent
+                                await eventsProcessor.process_event_ee(
+                                    distinct_id,
+                                    ip,
+                                    site_url,
+                                    processedEvent,
+                                    team_id,
+                                    DateTime.fromISO(now),
+                                    sent_at ? DateTime.fromISO(sent_at) : null
+                                )
+                                console.info(`Ingested event ${parsedEventMessage.data.event}!`)
+                            } else {
+                                console.info(`Discarding event ${parsedEventMessage.data.event}`)
+                            }
+                            server.statsd?.timing(`posthog-plugin-server-ingestion`, timer)
+                        }
+                        eventsProcessor.kafkaConsumer.commit()
```

Comment:
I think there's a timing issue here that may result in data loss. My knowledge of Kafka is limited, though I assume committing here moves some sort of "read index" down the list of events to what was just read, so they would not be read again. Basically I'm assuming it doesn't commit reads batch by batch, but just commits the last read index.. For example with batches A, C and D read and processed, batch B is still processing due to some long running code, committing D would also mark B as read.

Assuming a new batch is processed every 50ms, we can run into a situation where we get 100 messages in the first batch and 0 messages 50ms later, process them in an instant and commit "yup, we're here now", even though the first batch is still processing in the background.

Reply:
There is a potential issue, but it's actually different, as this is an at-least-once scenario – that is most events will be processed once, but occasionally more than once (so no data loss, the opposite). Exactly-once (like, _exactly_ exactly) consumption in Kafka is not a straightforward thing and requires transactions, which `node-rdkafka` doesn't support it seems.